### PR TITLE
A shared data structure was causing race conditions leading to VM creation with wrong configurations.

### PR DESCRIPTION
**Pull request by AxiomSamarth**
**What this PR does / why we need it**:
This PR isolates the providerSpec field and Secret field from the Driver object.

**Which issue(s) this PR fixes**:
Fixes [#304](https://github.com/gardener/gardener-extension-provider-azure/issues/304)

**Special notes for your reviewer**:

**Release note**:

```bugfix developer
A shared data structure was causing race conditions leading to VM creation with wrong configurations.
```

### CODE REVIEW

File: pkg/azure/core_test.go [hunk 1]
```diff
@@ -1683,7 +1673,7 @@ func UnmarshalInterfacesDeleteFuture(bytesUnterfacesFutureAPI []byte) network.In
 
 func newMachine(name string) *v1alpha1.Machine {
 	return &v1alpha1.Machine{
-		ObjectMeta: metav1.ObjectMeta{
+		ObjectMeta: v1.ObjectMeta{
```
Comment (ialidzhikov) [inline pkg/azure/core_test.go:1676]: You could revert this change. The idiomatic import alias for "k8s.io/apimachinery/pkg/apis/meta/v1" is `metav1`, not `v1`.

File: pkg/azure/utils.go [hunk 2]
```diff
@@ -234,50 +234,50 @@ func (d *MachinePlugin) getVMParameters(vmName string, image *compute.VirtualMac
...
 			},
 		}
 	}
 
 	return VMParameters
 }
 
-func getImageReference(d *MachinePlugin) compute.ImageReference {
-	imageRefClass := d.AzureProviderSpec.Properties.StorageProfile.ImageReference
+func (d *MachinePlugin) getImageReference(providerSpec *api.AzureProviderSpec) compute.ImageReference {
```
Comment (ialidzhikov) [inline pkg/azure/utils.go:279]: Why not?

```suggestion
func getImageReference(providerSpec *api.AzureProviderSpec) compute.ImageReference {
```

File: pkg/azure/utils.go [hunk 3]
```diff
@@ -157,21 +157,21 @@ func (d *MachinePlugin) generateDataDisks(vmName string, azureDataDisks []api.Az
 	return dataDisks
 }
 
-func (d *MachinePlugin) getVMParameters(vmName string, image *compute.VirtualMachineImage, networkInterfaceReferenceID string) compute.VirtualMachine {
+func (d *MachinePlugin) getVMParameters(vmName string, image *compute.VirtualMachineImage, networkInterfaceReferenceID string, providerSpec *api.AzureProviderSpec, secret *corev1.Secret) compute.VirtualMachine {
```
Comment (ialidzhikov) [inline pkg/azure/utils.go:160]: Does this func now need the MachinePlugin?

```suggestion
func getVMParameters(vmName string, image *compute.VirtualMachineImage, networkInterfaceReferenceID string, providerSpec *api.AzureProviderSpec, secret *corev1.Secret) compute.VirtualMachine {
```

File: pkg/azure/utils.go [hunk 4]
```diff
@@ -76,24 +77,23 @@ func getAzureDataDiskPrefix(name string, lun *int32) string {
...
-		var diskLun *int32
-		diskLun = disk.Lun
+		var diskLun *int32 = disk.Lun
 		azureDataDiskNames[i] = dependencyNameFromVMNameAndDependency(getAzureDataDiskPrefix(disk.Name, diskLun), vmname, suffix)
 	}
 	return azureDataDiskNames
 }
 
-func (d *MachinePlugin) getNICParameters(vmName string, subnet *network.Subnet) network.Interface {
+func (d *MachinePlugin) getNICParameters(vmName string, subnet *network.Subnet, providerSpec *api.AzureProviderSpec) network.Interface {
```
Comment (ialidzhikov) [inline pkg/azure/utils.go:86]:
```suggestion
func getNICParameters(vmName string, subnet *network.Subnet, providerSpec *api.AzureProviderSpec) network.Interface {Write<br class="Apple-interchange-newline">
```

**Review by prashanth26 — COMMENTED** [review]
For the future, avoid making beautification changes on hotfix PRs as sometimes this leads to other bugs. Preferably leave such changes for regular release cycle PRs.

**Review by ialidzhikov — APPROVED** [review]
/lgtm
/squash

File: pkg/azure/utils.go [hunk 5]
```diff
@@ -577,11 +576,9 @@ func (d *MachinePlugin) deleteVMNicDisks(ctx context.Context, clients spi.AzureD
 
 	deleters := []func() error{nicDeleter, diskDeleter}
 
-	if dataDiskNames != nil {
```
Comment (prashanth26) [inline pkg/azure/utils.go:580]: Why is this check removed? I hope this won't cause issues.

Comment (prashanth26) [reply pkg/azure/utils.go:580]: I think it's okay - https://github.com/gardener/machine-controller-manager-provider-azure/blob/master/pkg/azure/core.go#L144. Looks like it's never nil.

**Review by prashanth26 — APPROVED** [review]
/lgtm

File: pkg/azure/utils.go (hunk 5 above)
Comment (AxiomSamarth) [reply pkg/azure/utils.go:580]: If it is nil then the loop is never executed. No panics occur. So, this is good, @prashanth26 😄
